overlord/ifacestate/apparmorprompting/prompting.go

```
@@ -0,0 +1,551 @@
+		// Check that adding the rule doesn't conflict with other rules
+		newRule, err = m.rules.AddRule(userID, prompt.Snap, prompt.Interface, constraints, outcome, lifespan, duration)
+		if err != nil {
+			// Rule conflicts with existing rule (at least one identical pattern
+			// variant and permission). This should be considered a bad reply,
+			// since the user should only be prompted for permissions and paths
+			// which are not already covered.
+			return nil, err
+		}
+
+		defer func() {
+			if retErr != nil || lifespan == prompting.LifespanSingle {
+				m.rules.RemoveRule(userID, newRule.ID)
+			}
+		}()
+	}
+
+	prompt, retErr = m.prompts.Reply(userID, promptID, outcome)
+	if retErr != nil {
+		// Error should not occur unless the listener has closed
+		return nil, retErr
+	}
+
+	if lifespan == prompting.LifespanSingle {
+		return []prompting.IDType{}, nil
+	}
+
+	// Apply new rule to outstanding prompts.
+	satisfiedPromptIDs, err = m.applyRuleToOutstandingPrompts(newRule)
+	if err != nil {
+		// Should not occur, as outcome and constraints have already been
+		// validated. However, it's possible an error could occur if the prompt
+		// DB was already closed. This should be an internal error. However, we
+		// can't un-send any reply which might have been sent, and this should
+		// only fail if the prompting system is shutting down, so don't actually
+		// return an error.
+
+		// XXX: this log leaks information about internal activity
+		logger.Noticef("WARNING: error when handling new rule as a result of reply: %v", err)
+	}
+
+	return satisfiedPromptIDs, nil
+}
+
+func (m *InterfacesRequestsManager) applyRuleToOutstandingPrompts(rule *requestrules.Rule) (satisfiedPromptIDs []prompting.IDType, err error) {
+	metadata := &prompting.Metadata{
+		User:      rule.User,
+		Snap:      rule.Snap,
+		Interface: rule.Interface,
+	}
+	satisfiedPromptIDs, err = m.prompts.HandleNewRule(metadata, rule.Constraints, rule.Outcome)
+	return satisfiedPromptIDs, err
```

given that the callers are always ignoring errors maybe we can do it directly here? it avoid if and repeated code above

---

Really good idea, thank you!